lib/Backends/Interpreter/InterpreterFunction.cpp

```diff
@@ -229,21 +231,55 @@ Error BoundInterpreterFunction::execute(IRFunction *F,
     }
   }
 
-// Do the forward pass.
+  // Do the forward pass.
+  if (!getIRInstructionProcessingHandler()) {
+    // This is the fast version without hooks.
+    // Dispatch the interpreter on each instruction in the program:
+    for (const auto &I : F->getInstrs()) {
+      switch (I.getKind()) {
 #define DEF_VALUE(CLASS, NAME)
 #define DEF_INSTR(CLASS, NAME)                                                 \
   case Kinded::Kind::CLASS##Kind: {                                            \
     fwd##CLASS(llvm::cast<CLASS>(&I));                                         \
     break;                                                                     \
   }
 #define DEF_BACKEND_SPECIFIC_INSTR(CLASS, NAME)
-  // Dispatch the interpreter on each instruction in the program:
-  for (const auto &I : F->getInstrs()) {
-    switch (I.getKind()) {
 #include "glow/AutoGenInstr.def"
 
-    default:
-      llvm_unreachable("Invalid instruction.");
+      default:
+        glow::errs() << "Invalid instruction: " << &I << "\n";
+        llvm_unreachable("Invalid instruction.");
+      }
+    }
+  } else {
+    auto &irInstructionProcessingHandler = getIRInstructionProcessingHandler();
+    // Dispatch the interpreter on each instruction in the program:
+    for (const auto &I : F->getInstrs()) {
+      // Perform pre-processing and proceed with standard processing if needed.
+      if (!irInstructionProcessingHandler(
```

Comment:
Having PREPROCESSING and PROCESSING seems a bit redundant to me. If you want to skip the default implementation you can just return true from your custom handler and implement your  logic for that Instr there.

If you need to do some extra work before the default work, you can just return false for skip.

The way it is listed here it seems that you only do PROCESSING if your preprocessing returns false. Seems a bit redundant. WDYT?

Reply:
Yeah, I actually thought about it when I was defining processing stages.

But I guess my idea to explicitly distinguish between preprocessing, processing and post-processing comes from some past experiences with AOP (Aspect-oriented programming), where you explicitly have before/around/after advices.

Also, if we would allow for adding multiple hooks to the same backend, then you can get into a situation where you invoke a chain of hooks for the same instruction and the same processing stage and in that case you probably cannot always combine preprocessing with processing stages unless you are the last element of this chain, because the work done by these stages of the same hook would be separate in time (e.g. pre1 -> pre2 -> pre3 -> proc1 -> ...). Does it make sense?

Reply:
I'm a little confused here -- looks like if the handler returns true for PREPROCESSING here, then we skip PROCESSING entirely below? It seems like since there's no pre or post processing in the Interpreter currently then we would only need to check the handler's return from PROCESSING, to see if we skip the Interpreter's default processing.

Reply:
RE: multiple hooks, the current design doesn't allow for this, right? IMO it makes more sense to leave it without the extra stages since they aren't currently useful, and in the future we can add them + multiple hooks etc. as needed. Especially because the current design using an enum to pass the current stage is easily extended to handle these cases.

Reply:
@jfix71 

> It seems like since there's no pre or post processing in the Interpreter currently then we would only need to check the handler's return from PROCESSING, to see if we skip the Interpreter's default processing.

Oh, yes. I see your point. Will change.

Reply:
> IMO it makes more sense to leave it without the extra stages since they aren't currently useful, and in the future we can add them + multiple hooks etc. as needed. Especially because the current design using an enum to pass the current stage is easily extended to handle these cases.

So, I removed DONE. Do you propose to remove PREPROCESSSING as well like Sameer has proposed ?

Reply:
I probably would remove it since there's no clear use for it as currently formulated, but no strong feelings.

Reply:
OK, guys, you convinced me! :-) I removed PREPROCESSING.